src/kv/KeyValueDB.h

```diff
@@ -199,71 +199,78 @@ class KeyValueDB : public PriorityCache::PriCache {
     return get(prefix, string(key, keylen), value);
   }
 
-  // This superclass is used both by kv iterators *and* by the ObjectMap
-  // omap iterator.  The class hierarchies are unfortunately tied together
-  // by the legacy DBOjectMap implementation :(.
-  class SimplestIteratorImpl {
+  class ForwardIterator {
   public:
     virtual int seek_to_first() = 0;
-    virtual int upper_bound(const std::string &after) = 0;
-    virtual int lower_bound(const std::string &to) = 0;
     virtual bool valid() = 0;
-    virtual int next(bool validate=true) = 0;
+    virtual int next() = 0;
     virtual std::string key() = 0;
     virtual bufferlist value() = 0;
-    virtual int status() = 0;
-    virtual ~SimplestIteratorImpl() {}
-  };
-
-  class IteratorImpl : public SimplestIteratorImpl {
-  public:
-    virtual ~IteratorImpl() {}
-    virtual int seek_to_last() = 0;
-    virtual int prev(bool validate=true) = 0;
-    virtual std::pair<std::string, std::string> raw_key() = 0;
     virtual bufferptr value_as_ptr() {
       bufferlist bl = value();
       if (bl.length() == 1) {
         return *bl.buffers().begin();
       } else if (bl.length() == 0) {
         return bufferptr();
       } else {
-	ceph_abort();
+        ceph_abort();
       }
     }
+    virtual int status() = 0;
+    virtual ~ForwardIterator() {}
+  };
+
+  // This superclass is used both by kv iterators *and* by the ObjectMap
+  // omap iterator.  The class hierarchies are unfortunately tied together
+  // by the legacy DBOjectMap implementation :(.
```

Comment:
Lean toward either this impl (it's explicit) or the third one (single hierarchy, but we should make the backward iteration methods ceph_abort in default implementations).